### ISSUE TITLE
Incompatible checkpoint being restored on estimator with keras model

### ISSUE THREAD

**Issue by kurka** [tensorflow/tensorflow#24695]
**System information**
- Have I written custom code: yes
- OS Platform and Distribution: Archlinux
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.12.0
- Python version: 3.7.1
- CUDA/cuDNN version: 10.0.130-2/7.4.1.5-2
- GPU model and memory: 11GB GTX 1080 Ti

**Describe the current behavior**

I am running a Keras Model within a custom estimator and giving the size of a hidden layer as a parameter.
After each run with the estimator, a checkpoint is stored, saving the weights of that run. 

However, if the architecture of the model is changed between runs, instead of receiving an error due to incompatible checkpoints, tensorflow is loading the old checkpoint into the new model even though new dimension is being considered.

**Describe the expected behavior**

According to the [guide](https://www.tensorflow.org/guide/checkpoints#avoiding_a_bad_restoration), "restoring a model's state from a checkpoint only works if the model and checkpoint are compatible" and it is given an example of how the described situation (changes in the model architecture) should generate an `InvalidArgumentError`.

However no error is raised, being unclear the behaviour - are the new weights (in case of increased size) initialised and trained? Are they simply ignored?

**Code to reproduce the issue**

Below is a minimum working example to reproduce the problem:

```python
import tensorflow as tf
import numpy as np
from tensorflow.keras import layers


def main(argv):
    steps = 1000
    # Generate random data
    train_x = np.random.randn(100, 4).astype('float32')
    train_y = np.random.randint(0, 3, 100).astype('int32')

    # First estimator, with *10* hidden units
    classifier = tf.estimator.Estimator(
        model_fn=model_fn,
        model_dir="tmp",
        params={'hidden': 10}
    )

    # train first model and store checkpoints on "tmp"
    classifier.train(
        input_fn=lambda: tf.data.Dataset.from_tensors((train_x, train_y)),
        steps=steps)

    # second estimator, with *20* hidden units and same model_dir output
    # BUG: should break as model_dir is the same as previous estimator!!!
    classifier = tf.estimator.Estimator(
        model_fn=model_fn,
        model_dir="tmp",
        params={'hidden': 20}
    )

    classifier.train(
        input_fn=lambda: tf.data.Dataset.from_tensors((train_x, train_y)),
        steps=steps)


def model_fn(features, labels, mode, params):

    # create model with params[hidden] units on hidden layer
    model = tf.keras.Sequential([layers.Dense(params["hidden"], activation='relu'),
                                 layers.Dense(3)])
    logits = model(features)
    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
    optimizer = tf.train.AdagradOptimizer(learning_rate=0.1)
    train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())

    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)


if __name__ == '__main__':
    tf.logging.set_verbosity(tf.logging.INFO)
    tf.app.run(main)
```

**Other info / logs**
Log output (abbreviated)

~~~
# RUN 1
...
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into tmp/model.ckpt.
INFO:tensorflow:loss = 1.1404053, step = 0
INFO:tensorflow:Saving checkpoints for 1 into tmp/model.ckpt.
INFO:tensorflow:Loss for final step: 1.1404053.
...
# RUN 2
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.

INFO:tensorflow:Restoring parameters from tmp/model.ckpt-1
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1 into tmp/model.ckpt.
INFO:tensorflow:loss = 1.112431, step = 1
INFO:tensorflow:Saving checkpoints for 2 into tmp/model.ckpt.
INFO:tensorflow:Loss for final step: 1.112431.
~~~

**Comment by sanjoy** [comment]
Removing comp:gpu since this does not look like a GPU specific issue.  Please add it back if you disagree.

**Comment by allenlavoie** [comment]
Just an educated guess, but Keras Layers are numbered globally within the Python process. So if you re-create the Model you'll get totally new variable names each time. Then there's no conflict: it just looks like there are a bunch of new variables and the old ones don't match. Sad but a known incompatibility between Keras and name-based variable matching.

I'm somewhat more confused as to why Estimator isn't complaining or at least warning about variables that aren't in the checkpoint (given that it is restoring a checkpoint). Is that expected behavior? @k-w-w @karmel

**Comment by k-w-w** [comment]
All the checkpoint values are being loaded, but as a result the variable shapes are changed. 
It appears to be an unintentional bug from tf.train.Saver, not just a quirk of Estimator or Keras. The restore op doesn't do any shape checking. We should add a warning to saver.restore if any of the variables are reshaped.

**Comment by allenlavoie** [comment]
Assignment ops definitely do shape checking, though. Restores into variables with different shapes do fail.

**Comment by k-w-w** [comment]
tf.train.Saver uses RestoreV2/RestoreTensorsV2 instead of assign ops. Is this a bug with the restore ops?

**Comment by allenlavoie** [comment]
RestoreV2 just returns a Tensor with the value of the variable. Then that feeds into an assignment here: https://github.com/tensorflow/tensorflow/blob/55e68264ba670301fb861aa4d28f9642b251c6fc/tensorflow/python/training/saving/saveable_object_util.py#L113

There is a reshape, but in this case there'd be a different number of elements, so even if restored_shapes wasn't None (I think it is) that'd be an error.

**Comment by k-w-w** [comment]
I ran this and you're right, the restored_shapes is None. 
Looks like it is being set to None here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/saver.py#L336

So fixing this would just require creating savers with reshape=True.

**Comment by allenlavoie** [comment]
I don't think we want to reshape. shape_safe_assign_variable_handle does a shape check, so if we passed a restored_shapes argument we'd be doing less checking than we are now (exact shape match vs. only checking the number of elements). I've seen that error come up a bunch when refactoring models...

**Comment by allenlavoie** [comment]
Ah, I see. The shapes that come from RestoreV2 are completely undefined (None) so they're compatible with the variable's shape when in a tf.Graph().

Given that we always run RestoreV2 eagerly in 2.x except for Estimator, we should probably just make whichever tweak fixes this for Estimator. Maybe you're right that reshape=True is an easy workaround.

**Comment by David-LiCause** [comment]
@allenlavoie Do any of the TF releases include a fix for this issue?

**Comment by allenlavoie** [comment]
No, and AFAIK nobody is working on it. Since it only affects Estimator (in 2.x), maybe @karmel has someone who could take a look. I have a queue of other things that look like higher priorities at the moment.

Most of the work is going to be working to migrate existing code which breaks with new shape checking (or maybe we get lucky and it's fine). As @k-w-w mentioned, the Saver flag change itself is a pretty simple fix to try.

**Comment by David-LiCause** [comment]
@allenlavoie Thanks

**Comment by rmothukuru** [comment]
@kurka,
Documentation of [Estimators](https://www.tensorflow.org/guide/estimator#whats_supported_now) state:

> There is limited support for training with Estimator using all strategies except TPUStrategy. Basic training and evaluation should work, but a number of advanced features such as v1.train.Scaffold do not. There may also be a number of bugs in this integration and there are no plans to actively improve this support (the focus is on Keras and custom training loop support).

Can you please confirm if this issue is still relevant, as `Estimators` are not actively supported in `Tensorflow 2.x`? Thanks!
